File: repositories/langtags/language_repository.py
from core.sqlite.standard_table import StandardTable
from core.sqlite.standard_database import StandardDatabase
from utils.datetime_util import get_datetime_now, set_datetime_formatted

class LanguageRepository:
    def __init__(self, table: StandardTable):
        self.table = table
        self.database = self.table.database

    def update_code(self, language_id:int, code: str):
        try:
            cursor = self.database.execute(
                statement=f"UPDATE {self.table.name} SET code=?, updated_at=? WHERE language_id=?;", commit=True, params=(code,get_datetime_now(),language_id)
            )
            return True
        except:
            return False

    def insert_code(self, code:str):
        try:
            cursor = self.database.execute(
                statement=(
                    f"INSERT INTO {self.table.name} (code,created_at) VALUES(?, ?);"
                ),
                commit=True, params=(code,get_datetime_now())
            )
            return True
        except:
            return False

    def code_exists(self, language_id:int, code: str) -> bool:
        try:
            cursor = self.database.execute(
                statement=f'SELECT 1 FROM {self.table.name} WHERE language_id=? AND code=? LIMIT 1;',
                commit=False, params=(language_id,code)
            )
            return cursor.fetchone() is not None
        except:
            return False

    def get_code_id(self, code:str) -> int | None:
        try:
            cursor = self.database.execute(
                statement=f"SELECT language_id FROM {self.table.name} WHERE code=? LIMIT 1;",
                commit=False, params=(code,)
            )
            row = cursor.fetchone()
            return row[0] if row else None
        except:
            return None

    def save_code(self, code:str):
        language_id = self.get_code_id(code)
        updated = False
        inserted = False
        if language_id is not None:
            updated = self.update_code( language_id, code )
        if updated == False:
            inserted = self.insert_code( code )
        return updated or inserted


    def deactivate(self, language_id: int):
        try:
            cursor = self.database.execute(
                statement=f"UPDATE {self.table.name} SET deleted_at=? WHERE language_id=?;",
                commit=True, params=(get_datetime_now(), language_id)
            )
            return True
        except:
            return False

    def activate(self, language_id: int):
        try:
            cursor = self.database.execute(
                statement=f"UPDATE {self.table.name} SET deleted_at=? WHERE language_id=?;",
                commit=True, params=(None, language_id)
            )
            return True
        except:
            return False

    def is_deleted(self, language_id:str) -> bool:
        try:
            cursor = self.database.execute(
                statement=f"SELECT deleted_at FROM {self.table.name} WHERE language_id=? LIMIT 1;",
                commit=False, params=(language_id,)
            )
            row = cursor.fetchone()
            return row[0] is not None
        except:
            return False

    def toggle_code_state(self, code: str):
        language_id = self.get_code_id(code)
        if language_id is not None:
            deleted = self.is_deleted( language_id )
            if deleted:
                return self.activate( language_id )
            else:
                return self.deactivate( language_id )
        return False

File: Repositories/LangTags/LanguageRepository.cs
// SQLite
using Microsoft.Data.Sqlite;

// Entidades
using Entities.LangTags;

// Core
using Core.Sqlite;

// Utils
using Utils.Text;

namespace Repositories.LangTags {
    public class LanguageRepository {
        // Variables
        private StandardDatabase _db;
        private StandardTable _table;

        // Constructor
        public LanguageRepository(StandardDatabase standardDatabase){
            _db = standardDatabase;
            _table = new StandardTable(_db, "languages");
        }

        // Propiedades
        public StandardTable Table { get {return _table;} }

        // Methods. No TRY, full crash si pasas cosas mal.
        public void Insert(string code, bool isActive){     
            // Solo create at y deleted at para insert. updated at solo update method.
            string now = TextualDateTime.ToText(DateTime.Now);
            object deletedAt = isActive ? DBNull.Value : now;
            _db.Execute(
                sql: "INSERT INTO languages (code, created_at, updated_at, deleted_at, is_active) VALUES (@p0, @p1, NULL, @p2, @p3);", commit: true,
                code, now, deletedAt, isActive ? 1 : 0
            );
        }

        public void Update(int languageId, string code, bool isActive )
        {
            // Solo update at si esta active.
            string now = TextualDateTime.ToText(DateTime.Now);
            object updatedAt = isActive ? now : DBNull.Value;
            object deletedAt = isActive ? DBNull.Value : now;
            _db.Execute(
                sql: "UPDATE languages SET code=@p0, updated_at=@p1, deleted_at=@p2, is_active=@p3 WHERE language_id=@p4;", commit:true,
                code, updatedAt, deletedAt, isActive ? 1 : 0, languageId
            );
        }

        public bool ExistsById(int languageId){
            using var reader = _db.Query(
                sql: "SELECT 1 FROM languages WHERE language_id=@p0 LIMIT 1;",
                languageId
            );
            return reader.Read();
        }
        public bool ExistsByCode(string code){
            using var reader = _db.Query(
                sql: "SELECT 1 FROM languages WHERE code=@p0 LIMIT 1;",
                code
            );
            return reader.Read();
        }

        public int GetIdByCode(string code){
            using var reader = _db.Query(
                sql: "SELECT language_id FROM languages WHERE code=@p0 LIMIT 1;",
                code
            );
            reader.Read();
            return reader.GetInt32(0);
        }

        public string GetCodeById(int languageId){
            using var reader = _db.Query(
                sql: "SELECT code FROM languages WHERE language_id=@p0 LIMIT 1;",
                languageId
            );
            reader.Read();
            return reader.GetString(0);
        }

        public void Save(string code, bool isActive=true, int? languageId=null)
        {
            // Determinar que exista
            bool exists = false;
            int id = 0;
            if (languageId == null) {
                exists = ExistsByCode(code);
                if (exists){
                    id = GetIdByCode(code);
                }
            } else {
                id = languageId.Value;
                exists = ExistsById(id);
            }
            
            // Escribir data.
            if (exists) {
                Update(id, code, isActive);
            } else {
                Insert(code, isActive);
            }
        }

        public void Activate(int languageId)
        {
            string now = TextualDateTime.ToText(DateTime.Now);
            _db.Execute(
                sql: "UPDATE languages SET updated_at=@p0, deleted_at=NULL, is_active=1 WHERE language_id=@p1;", commit:true,
                now, languageId
            );
        }
        public void Deactivate(int languageId)
        {
            string now = TextualDateTime.ToText(DateTime.Now);
            _db.Execute(
                sql: "UPDATE languages SET deleted_at=@p0, is_active=0 WHERE language_id=@p1;", commit:true,
                now, languageId
            );
        }
        //
    }
}